apps/walletapi/lib/walletapi/resolver/currency_conversion/exchange_rate_api.ex

```diff
@@ -0,0 +1,94 @@
+defmodule WalletApi.CurrencyConversion.ExchangeRateAPI do
+  @moduledoc """
+   get exchange rate from currency layer API
+  """
+  alias WalletApi.DisplayError
+  @api_url Application.fetch_env!(:walletapi, :exchange_rates_api)
+  # behavior added to mock query_exchange_rate() for testing.
+  @behaviour WalletAPI.Resolver.CurrencyConversion.ExchangeRateBehaviour
+  require Logger
+
+  def get_exchange_rate_behaviour, do: Application.get_env(:walletapi, :exchange_rate_behaviour)
+
+  defp get_quotes({:error, :not_found}, _timestamp), do: {:error, :not_found}
+
+  defp get_quotes(response, pair) do
+    case response["quotes"][pair] do
+      nil -> {:error, :not_found}
+      _ -> Decimal.from_float(response["quotes"][pair])
+    end
+  end
+
+  def get_exchange_rate(source_currency_code, currency_code, timestamp) do
+    if(!currency_code) do
+      raise DisplayError, message: "No currency code specified"
+    end
+
+    pair = source_currency_code <> currency_code
+    response = get_exchange_rate_behaviour().query_exchange_rate(source_currency_code, currency_code, timestamp)
+    get_quotes(response, pair)
+  end
+
+  def query_exchange_rate(source_currency_code, _currency_code, timestamp) do
+    time_in_ms =
+      if timestamp === nil do
+        DateTime.utc_now()
+      else
+        DateTime.from_unix!(timestamp, :millisecond)
+      end
+
+    # is08601 format : YYYY-MM-DD
+    date = Date.to_iso8601(time_in_ms)
+
+    # if given time within last 24 hours - get current exchange rates from API.
+    if DateTime.diff(DateTime.utc_now(), time_in_ms) < 24 * 60 * 60 do
```

Comment:
Hmm, does it mean that within the current 24 hours time frame every request will result in an API request?

Reply:
Yes, took the reference from blockchain-api. We are fetching the recent exchange_rates for transaction that happened within the last 24hrs.

Reply:
But if I'm reading your code correctly, if the time is within 24 hours, you'll make an API call every time to fetch the rate, unlike blockchain API code that would make a call once and cache it for 12 hours (https://github.com/celo-org/celo-monorepo/blob/master/packages/blockchain-api/src/currencyConversion/ExchangeRateAPI.ts#L47). Am I missing something? Will ConCache do it behind the scenes as well? It's a bit implicit, so hard to say what would be the behavior in this case.

Reply:
Modified the code to save the exchange rates for 12 hours for queries with time less than 24hrs.